File: lib/rest/yampi/services/yampi_marketing_service.dart
import 'package:sertton/core/global/responses/rest_response.dart';
import 'package:sertton/core/marketing/dtos/banner_dto.dart';
import 'package:sertton/core/marketing/dtos/lead_dto.dart';
import 'package:sertton/core/marketing/interfaces/marketing_service.dart';
import 'package:sertton/rest/yampi/mappers/yampi_banner_mapper.dart';
import 'package:sertton/rest/yampi/services/yampi_service.dart';

class YampiMarketingService extends YampiService implements MarketingService {
  YampiMarketingService(super.restClient, super.envDriver);

  @override
  Future<RestResponse<List<BannerDto>>> fetchBanners() async {
    final response = await super.restClient.get('/marketing/banners');
    return response.mapBody((body) {
      if (response.isFailure) return [];
      return YampiBannerMapper.toDtoList(body);
    });
  }

  @override
  Future<RestResponse<void>> saveLead(LeadDto leadDto) {
    throw UnimplementedError();
  }
}

File: src/rest/yampi/services/yampi-marketing-service.ts
import type { IMarketingService } from "@/core/marketing/interfaces/marketing-service"
import type { RestClient } from "@/core/shared/interfaces/rest-client"
import { YampiBannerMapper } from "@/rest/yampi/mappers"
import type { YampiBanner, YampiResponse } from "@/rest/yampi/types"

export const YampiMarketingService = (restClient: RestClient): IMarketingService => {
  const bannerMapper = YampiBannerMapper()

  return {
    async fetchBanners() {
      const response = await restClient.get<YampiResponse<YampiBanner>>("/marketing/banners")
      return response.mapBody((body) => body.data.map(bannerMapper.toDomain))
    },

    async saveLead(lead) {
      const response = await restClient.post<void>("/leads", {
        body: { email: lead.email, ...(lead.name ? { name: lead.name } : {}) },
      })
      return response
    },
  }
}
